### ISSUE TITLE
Video Data Preprocessing Takes Too Long

### ISSUE THREAD

**Issue by JacksonHuang127** [facebookresearch/perception_models#26]
Hi,

First of all, thank you for sharing your excellent work. I’ve encountered an issue while performing SFT on the PLM model where the video data preprocessing is taking an unexpectedly long time.

I am using a dataset with a total of 5,000 entries, where:
4,000 are images
1,000 are videos

My dataset format:
![Image](https://github.com/user-attachments/assets/5cfa4e1c-9bed-46f7-b1e2-4d06d681925e)

During the SFT process, the program spends an extended period processing video frames, and the preprocessing time is significantly longer than expected. 

Here is the output I am seeing:

![Image](https://github.com/user-attachments/assets/bf7dfef1-2249-4bbc-9c17-2a07d65aad5f)

Is there any advice on how to speed up the video frame preprocessing, or is there an option I might have missed to optimize this part of the process?

Thanks in advance for your help!

**Comment by hlh2023214** [comment]
I tried using the official dataset, and it runs fine, but when I switch to my own dataset, it keeps repeating:

![Image](https://github.com/user-attachments/assets/3ecef183-f8f0-4e79-9de3-fc87b2f99f4f)

This is the path to my dataset:

![Image](https://github.com/user-attachments/assets/d06595ea-956f-46f3-ad42-9addaa2b2112)

![Image](https://github.com/user-attachments/assets/91a43b7c-8ea0-46f6-8aa2-cb2e58afac8c)

**Comment by mmaaz60** [comment]
Hi @hlh2023214,

Thank you for your interest in our work. If I correctly understood, the SFT is taking significantly more time when using your custom data as compared to when using the provided dummy data.

- If that is the case, I would recommend looking at your videos for any errors, for example some videos might be corrupt. Its unlikely that this is the issue, but it is better to double check.

- Also, please try to decrease number of workers and run on a single node, as it might be helpful for small datasets (i.e. of size approx. 5K). Further, try reducing the logging freq.

In case if the issue persists, please share the training config that you are using, along with a few samples (if possible) from the dataset to reproduce the issue.

Thanks.

**Comment by hlh2023214** [comment]
Thank you for your reply. I followed the process outlined in apps/plm/docs/finetune_example.md to configure my own dataset and my data is all normal. I would like to know if this warning will affect my training process. Is it correctly loading the videos or images? I don't fully understand what this means:

**Setting JSONL position on /apdcephfs_sz/private_jkshuang/PLM-data/PLM_plm.jsonl (6/16): None.**

Is it because this is the first training on this data and it cannot find the resume?
In my previous tests, I only used the official image dataset, and the same issue occurs when loading the video dataset as well.
![Image](https://github.com/user-attachments/assets/a88a94ac-17dd-4598-adf0-cc74ad6e6b8a)

**Comment by mmaaz60** [comment]
Hi @hlh2023214,

This warning means that you are training for more iterations than the number of samples in your `jsonl` file. Please make sure that `steps: xxx` matches the number of samples you have in your jsonl to avoid this warning. For example, `steps = #gpus / #batch_size`.

If you want to train for multiple-epochs, please repeat the samples in your `jsonl` file and adjust the steps accordingly.

It is recommended to create `jsonl` containing multiple epochs of the data where you may shuffle lines for each epoch.

However, please note that this warning does not effect the training. The code automatically restart the iteration from the start of the `jsonl` if the data finishes.

I hope this will help. Thank You.


https://github.com/facebookresearch/perception_models/blob/f6372458272f5902841cb07b13da293949b8e4dd/core/data/data_mixer.py#L73

**Comment by hlh2023214** [comment]
Thank you for your detailed reply. I understand why each piece of data in the dummy dataset's **annotation.jsonl** is repeated multiple times.  

I also have a few other questions:  
1. To improve the model's precision in **video temporal localization** tasks, can I reduce the image size to **224** to increase the frame sampling rate and maximum number of frames?  

2. During **multi-object detection** in images, the model tends to output duplicate predictions. Is this due to an excessive number of tokens? Are there any optimization suggestions? 

question = "Locate every item from the category list in the image and output the coordinates in JSON format. The category set includes person, bicycle, car, motorcycle, airplane, bus, train, truck, boat, traffic light, fire hydrant, stop sign, parking meter, bench, bird, cat, dog, horse, sheep, cow, elephant, bear, zebra, giraffe, backpack, umbrella, handbag, tie, suitcase, frisbee, skis, snowboard, sports ball, kite, baseball bat, baseball glove, skateboard, surfboard, tennis racket, bottle, wine glass, cup, fork, knife, spoon, bowl, banana, apple, sandwich, orange, broccoli, carrot, hot dog, pizza, donut, cake, chair, couch, potted plant, bed, dining table, toilet, tv, laptop, mouse, remote, keyboard, cell phone, microwave, oven, toaster, sink, refrigerator, book, clock, vase, scissors, teddy bear, hair drier, toothbrush."

output：
[{'umbrella': [57, 71, 201, 231]}, {'umbrella': [201, 83, 334, 211]}, {'umbrella': [346, 102, 598, 231]}, {'umbrella': [270, 231, 423, 371]}, {'umbrella': [463, 96, 605, 231]}, {'person': [97, 102, 201, 427]}, {'person': [214, 136, 283, 427]}, **{'person': [302, 147, 376, 427]}, {'person': [386, 147, 464, 427]}, {'person': [463, 147, 532, 427]}, {'person': [463, 147, 530, 427]}, {'person': [400, 147, 464, 427]}, {'person': [302, 147, 376, 427]}, {'person': [400, 147, 464, 427]}, {'person': [463, 147, 510, 427]}, {'car': [38, 320, 110, 350]}, {'car': [38, 320,110, 350]}**

**Comment by mmaaz60** [comment]
Hi @hlh2023214,

1. Yes, to afford more frames you can either use low resolution of try to pool the features more aggressively. Please note that we did not test either of the configuration, but would be worth exploring. Do let me know you observe any interesting behavior.
2. PLM was not trained to work for multi-object detection, so additional NMS will be required. We recommend to use for grounding (single box) per category.

Thank You.
